filebeat/beater/filebeat.go

```diff
@@ -96,23 +105,47 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	defer spooler.Stop()
 	defer publisherChan.Close()
 
-	err = crawler.Start(registrar.GetStates())
+	err = crawler.Start(registrar.GetStates(), *once)
 	if err != nil {
 		return err
 	}
 	// Stop crawler -> stop prospectors -> stop harvesters
 	defer crawler.Stop()
 
-	// Blocks progressing. As soon as channel is closed, all defer statements come into play
+	// Wait for completion if running once.
+	if *once {
+		go func() {
+			logp.Info("Running filebeat once. Waiting for completion ...")
+			crawler.WaitForCompletion()
+
+			// Flushes spooler as often as needed to send all events to publisher
+			for spooler.Flush() > 0 {
+			}
+
+			// Wait for registrar to finish writing registry
+			finishedWg.Wait()
+			logp.Info("All data collection completed. Shutting down.")
```

Comment:
I wonder if `finishedWG.Wait()` is better done via defer statement by modifying the spooler shutdown order in here to:

```
if *once {
    defer publisherChan.Close() // close publisher channel (optional)
    defer finishedWG.Wait() // wait for all events processed by registrar
    defer spooler.Stop() // stop + flush spooler
} else {
    defer spooler.Stop() // stop spooler
    defer publisherChan.Close() // close publisher channel potentially unblocking the spooler
}
```